pillarbox-demo/src/main/java/ch/srgssr/pillarbox/demo/ui/showcases/layouts/OptimizedStory.kt

```diff
@@ -4,112 +4,184 @@
+ * Optimized story-like layout.
  */
 @Composable
 fun OptimizedStory(storyViewModel: StoryViewModel = viewModel()) {
-    val pagerState = rememberPagerState {
-        storyViewModel.playlist.items.size
+    val mediaItems = storyViewModel.mediaItems
+    val pagerState = rememberPagerState { mediaItems.size }
+    val settledPage by remember { derivedStateOf { pagerState.settledPage } }
+    LaunchedEffect(settledPage) {
+        storyViewModel.setActivePage(settledPage)
     }
-    LifecycleStartEffect(pagerState) {
-        storyViewModel.getPlayerForPageNumber(pagerState.currentPage).play()
 
-        onStopOrDispose {
-            storyViewModel.pauseAllPlayer()
-        }
-    }
-
-    val playlist = storyViewModel.playlist.items
     Box(modifier = Modifier.fillMaxSize()) {
-        HorizontalPager(
-            beyondViewportPageCount = 0,
-            key = { page -> playlist[page].uri },
+        VerticalPager(
             flingBehavior = PagerDefaults.flingBehavior(
                 state = pagerState,
                 pagerSnapDistance = PagerSnapDistance.atMost(0),
-                snapAnimationSpec = spring(stiffness = Spring.StiffnessHigh)
             ),
-            pageSpacing = 1.dp,
-            state = pagerState
+            beyondViewportPageCount = 0,
+            state = pagerState,
         ) { page ->
-            // When flinging -> may "load" more that 3 pages
-            val currentPage = pagerState.currentPage
-            val player = if (page == currentPage - 1 || page == currentPage + 1 || page == currentPage) {
-                val playerConfig = storyViewModel.getPlayerAndMediaItemIndexForPage(page)
-                val playerPage = storyViewModel.getPlayerFromIndex(playerConfig.first)
-                playerPage.playWhenReady = currentPage == page
-                playerPage.seekToDefaultPosition(playerConfig.second)
-                playerPage
-            } else {
-                null
-            }
-            player?.let {
+            val player = remember { storyViewModel.getConfiguredPlayerForPageNumber(page) }
+
+            val progress by remember {
+                player.currentPositionAsFlow(100.milliseconds)
+                    .map { it / player.duration.coerceAtLeast(1L).toFloat() }
+            }.collectAsState(0f)
+
+            val isBuffering by remember {
+                player.playbackStateAsFlow().map { it == Player.STATE_BUFFERING }
+            }.collectAsState(false)
+
+            Box(
+                modifier = Modifier.fillMaxSize(),
+            ) {
                 PlayerSurface(
                     modifier = Modifier.fillMaxHeight(),
                     scaleMode = ScaleMode.Crop,
-                    // Using Texture instead of Surface because on Android API 34 animations are not working well due to the hack
-                    // See PlayerSurfaceView in AndroidPlayerSurfaceView
                     surfaceType = SurfaceType.Texture,
```

Comment:
Since we know that the issue is only on API 34, should we do something smarter here? Like:
```suggestion
                    surfaceType = if (Build.VERSION.SDK_INT == Build.VERSION_CODES.UPSIDE_DOWN_CAKE) SurfaceType.Texture else SurfaceType.Surface,
```